### PR TITLE
Anthonyw298

### PR DESCRIPTION
**Pull request by anthonyw298**
Changed layout of Base Code where I took away fans and kept the same files as main branch. Only changed the pump variables by adding a structure type

### CODE REVIEW

File: Simulation Code/Base/Base.ino [hunk 1]
```diff
@@ -1,112 +1,142 @@
...
-    {
-      PumpPWM = 170; // Set the pump speed to medium-high
-      previousTime = currentTime; // Update the time of the last water level check
-    }
-    else if ((waterLevelC == HIGH) && (elapsedTime >= stabilityTime))
-    {
-      PumpPWM = 85; // Set the pump speed to medium-low
-      previousTime = currentTime; // Update the time of the last water level check
+// Initialize the fans
+void beginFans(Fans &fans) {
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:62]: dont create this function

File: Simulation Code/Base/Base.ino [hunk 2]
```diff
@@ -1,112 +1,142 @@
...
+    pinMode(fans.pinA, OUTPUT);
+    pinMode(fans.pinB, OUTPUT);
+    fans.previousTime = millis();
+}
+
+// Update fan control based on time
+void updateFans(Fans &fans, unsigned long currentTime) {
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:69]: dont create this function

File: Simulation Code/Base/Base.ino [hunk 3]
```diff
@@ -1,112 +1,142 @@
...
 
-// Function to check the current water level
-void checkWaterLevel(int &waterLevelA, int &waterLevelB, int &waterLevelC) 
-{
-  // Read the value of the water level sensors
-  waterLevelA = digitalRead(WaterLevelA);
-  waterLevelB = digitalRead(WaterLevelB);
-  waterLevelC = digitalRead(WaterLevelC);
+// Read water levels from sensors
+void readWaterLevels(WaterLevelSensors &sensors, int &levelA, int &levelB, int &levelC) {
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:55]: dont create this function

File: Simulation Code/Base/Base.ino [hunk 4]
```diff
@@ -1,112 +1,142 @@
...
+    const unsigned long pumpTimeOff;
+    const unsigned long pumpTimeOn;
+
+    WaterPumpController(int p)
+        : pin(p), pumpPWM(0), pumpStatus(false), previousTime(0), previousPumpTime(0),
+          stabilityTime(1000), pumpTimeOff(1000), pumpTimeOn(5000) {}
+};
+
+// Initialize the water level sensors
+void beginWaterLevelSensors(WaterLevelSensors &sensors) {
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:48]: dont create this function

File: Simulation Code/Base/Base.ino [hunk 5]
```diff
@@ -1,112 +1,142 @@
...
+    int pinA;
+    int pinB;
+    unsigned long previousTime;
+    int state;
+
+    Fans(int aPin, int bPin) : pinA(aPin), pinB(bPin), previousTime(0), state(0) {}
+};
+
+// Define a structure for water pump control
+struct WaterPumpController {
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:32]: this is how I want a structure to look

File: Simulation Code/Base/Base.ino [hunk 6]
```diff
@@ -1,112 +1,142 @@
...
+    int pin;
+    int pumpPWM;
+    bool pumpStatus;
+    unsigned long previousTime;
+    unsigned long previousPumpTime;
+    const unsigned long stabilityTime;
+    const unsigned long pumpTimeOff;
+    const unsigned long pumpTimeOn;
+
+    WaterPumpController(int p)
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:42]: dont initialize inside the struct however

File: Simulation Code/Base/Base.ino [hunk 7]
```diff
@@ -1,112 +1,142 @@
...
+};
+
+// Define a structure for fan control
+struct Fans {
+    int pinA;
+    int pinB;
+    unsigned long previousTime;
+    int state;
+
+    Fans(int aPin, int bPin) : pinA(aPin), pinB(bPin), previousTime(0), state(0) {}
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:28]: dont initialize inside the structure

File: Simulation Code/Base/Base.ino [hunk 8]
```diff
@@ -1,112 +1,142 @@
...
-  pinMode(WaterPump, OUTPUT);
+const int WaterPump = 11;
+
+// Define a structure for water level sensors
+struct WaterLevelSensors {
+    int pinA;
+    int pinB;
+    int pinC;
+
+    WaterLevelSensors(int a, int b, int c) : pinA(a), pinB(b), pinC(c) {}
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:18]: dont initialize inside the structure, initialization occurs in void setup

File: Simulation Code/Base/Base.ino [hunk 9]
```diff
@@ -1,112 +1,142 @@
...
-unsigned long previousTime = 0; // Variable to store the last time the water level was checked
-unsigned long previousPumpTime = 0; // Variable to store the last time the pump status was checked
-const unsigned long stabilityTime = 1000; // Time in milliseconds to wait before changing the pump speed
-bool pumpStatus = false; // Variable to store the current status of the pump (true = on, false = off)
-const unsigned long pumpTimeOff = 1000; // Time in milliseconds for the pump to stay off
-const unsigned long pumpTimeOn = 5000; // Time in milliseconds for the pump to stay on
-const int WaterPump = 11; // Pin for the water pump
-int PumpPWM; // Variable to control the speed of the water pump
-
-void setup()
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:24]: this is how I want void setup to be, don't initialize inside structures

File: Simulation Code/Base/Base.ino [hunk 10]
```diff
@@ -1,5 +1,5 @@
 #include <Arduino.h>
-
+//Actual Revised Code
 // Define the pins for the water level sensors
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:3]: Correct me if I am wrong but I think we should group all the water level sensors together in a struct. So both the pin declarations and the variables that stores the water levels can be in the same struct.

File: Simulation Code/Base/Base.ino [hunk 11]
```diff
@@ -11,15 +11,22 @@ const int FanA = 9;
 const int FanB = 10;
 int FanPWM; // Variable to control the speed of the fans
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:12]: Same thing with the fans, group into struct and rename

File: Simulation Code/Base/Base.ino [hunk 12]
```diff
@@ -11,15 +11,22 @@ const int FanA = 9;
...
+
+    const unsigned long stabilityTime = 1000; // Stability time in milliseconds
+    const unsigned long pumpTimeOff = 1000; // Pump off time in milliseconds
+    const unsigned long pumpTimeOn = 5000; // Pump on time in milliseconds
+};
+
+// Create an instance of the WaterPumpFn struct
+WaterPumpFn waterPump;
+
 const int WaterPump = 11; // Pin for the water pump
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:29]: Lets throw this in the struct

File: Simulation Code/Base/Base.ino (hunk 10 above)
Comment (david1534) [reply Simulation Code/Base/Base.ino:3]: We should also rename any pin declarations from WaterLevelA to WaterPinA format so it is easier to tell between variables and pins

File: Simulation Code/Base/Base.ino [hunk 13]
```diff
@@ -106,7 +113,6 @@ void loop()
   unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:113]: Can these variables all be grouped with their respective struct? maybe we can simply create an instance in the loop if needed

File: Simulation Code/Base/Base.ino [hunk 14]
```diff
@@ -106,7 +113,6 @@ void loop()
...
 
   checkWaterLevel(waterLevelA, waterLevelB, waterLevelC); // Check the current water level
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:115]: The point of the struct is to have it passed into functions all at once. For example, checkWaterLevel(WaterLevelStruct);

File: Simulation Code/Base/Base.ino [hunk 15]
```diff
@@ -106,7 +113,6 @@ void loop()
...
-  checkPumpStatus(currentTime, elapsedTime, currentPumpTime, elapsedPumpTime, 
+  checkPumpStatus(waterPump, currentTime, elapsedTime, currentPumpTime, elapsedPumpTime, 
   waterLevelA, waterLevelB, waterLevelC); // Check and update the pump status
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:117]: Same thing here

File: Simulation Code/Base/Base.ino [hunk 16]
```diff
@@ -44,59 +51,59 @@ void checkWaterLevel(int &waterLevelA, int &waterLevelB, int &waterLevelC)
 }
 
 // Function to check and update the status of the pump
-void checkPumpStatus(unsigned long &currentTime, unsigned long &elapsedTime, 
+void checkPumpStatus(WaterPumpFn &waterPump, unsigned long &currentTime, unsigned long &elapsedTime,
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:54]: See comment at the bottom, but the point of a struct is to group similar variables or pin declarations together

**Review by david1534 — CHANGES_REQUESTED** [review]
Let me know if these changes are feasible to implement. I'm not 100% sure if it is so I wanted you to explore the proposed changes and validate it with the electrical simulation before you update this pull request. Do not delete this current request, I will resolve the comments in this review when the necessary changes are addressed or corrected.

File: Simulation Code/Base/Base.ino [hunk 17]
```diff
@@ -2,46 +2,63 @@
...
+  unsigned long previousTime = 0;
+  unsigned long previousPumpTime = 0;
+  const unsigned long stabilityTime = 1000;
+  bool pumpStatus = false;
+  const unsigned long pumpTimeOff = 1000;
+  const unsigned long pumpTimeOn = 5000;
+  const int WaterPump = 11;
+  int PumpPWM;
+  unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime;
+};
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:34]: Is it possible to have these structures defined in its own header file. One specifically for structures?

File: Simulation Code/Base/Base.ino [hunk 18]
```diff
@@ -2,46 +2,63 @@
...
+  Serial.begin(9600);
+
+  // Set the mode of the pins using the struct instances
+  pinMode(waterLevelSensors.WaterPinA, INPUT);
+  pinMode(waterLevelSensors.WaterPinB, INPUT);
+  pinMode(waterLevelSensors.WaterPinC, INPUT);
+  pinMode(waterLevelSensors.WaterPinD, OUTPUT);
+  pinMode(fans.FanA, OUTPUT);
+  pinMode(fans.FanB, OUTPUT);
+  pinMode(waterPumpFn.WaterPump, OUTPUT);
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:52]: Good use of structure

File: Simulation Code/Base/Base.ino [hunk 19]
```diff
@@ -2,46 +2,63 @@
...
 }
 
 void loop() 
 {
-  int waterLevelA, waterLevelB, waterLevelC, waterLevelD; // Variables to store the current water level
-  unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
+  waterPumpFn.currentTime = millis();
```
Comment (david1534) [inline Simulation Code/Base/Base.ino:57]: I'm assuming this has to be declared within the loop?

File: Simulation Code/Base/README.txt
Comment (david1534) [inline Simulation Code/Base/README.txt:1]: Add a readme file for descriptions of each of the files within this folder

**Review by david1534 — CHANGES_REQUESTED** [review]
This is good and basically ready for merging. See comments @anthonyw298

File: Simulation Code/Base/Base.ino (hunk 19 above)
Comment (anthonyw298) [reply Simulation Code/Base/Base.ino:57]: yeah since its going to be constantly running

File: Simulation Code/Base/Base.ino (hunk 17 above)
Comment (anthonyw298) [reply Simulation Code/Base/Base.ino:34]: not sure what you mean by this

**Review by david1534 — APPROVED** [review]

**Pull request merged**